### PR TITLE
Do not skip connections in bidiastar

### PR DESCRIPTION
**Pull request by genadz**
# Issue

This PR should fix 2 bugs:
**1) for some cases bidirastar builds extra detours.**

After debugging:
in bidirectional astar: we mark edge as `permanent` only before call expansion method. But it's not correct, because we can mark edge earlier right after we popped this edge from the queue.

Current behavior may loose some connections in the following case: let's suppose `A->B->C->D` is the part of the shortest route. On some iteration we popped `B->C` label as `fwd_pred` from the forward queue. Then we applied reverse search from the node `C` (suppose cost for `D->C` less than cost for `B->C`). Then we popped `C->B` label as `rev_pred` from the reverse queue. And, looks like now we should set the connection `B->C`, but we don't do it because the edge `B->C` hasn't marked as `permanent` for now. So, we skip this and continue forward expansion from the node `C`.

So, now imagine that we have 2 not-thru regions `A` and `B` and one connection edge between these regions. In some cases if we skip this connection path cannot be found or may be found with extra detour.

**2) For some cases bidirastar returns pathes with double u-turns.**

It happens due to the following: let's suppose we have a complex restriction `no_u_turn: from AB via BC to CD`. Then, forward search reaches `BC`, checks that it's a deadend (because of the restriction) and makes a u-turn `CB`. After that reverse search reaches `CB`, checks that it's a deadend (because of the restriction), makes a u-turn `BC` and meets `CB` edge from forward search. So, now we have the following path: `AB -> BC -> CB -> BC -> CD` - there is no rule to forbid such path so we accept it. 

## Tasklist

 - [x] Add tests
 - [x] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [x] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by genadz** [comment]
When I made changes some tests failed. The reason: we have at least one more bug. This bug should be fixed here https://github.com/valhalla/valhalla/pull/2796 . So, this PR is blocked right now

**Comment by genadz** [comment]
RAD tests passed. It was only one different route, I checked it manually, no worries.

File: src/thor/bidirectional_astar.cc [hunk 1]
```diff
@@ -758,7 +758,7 @@ bool BidirectionalAStar::SetForwardConnection(GraphReader& graphreader, const BD
 
   // Set a threshold to extend search
   if (threshold_ == std::numeric_limits<float>::max()) {
-    threshold_ = (pred.sortcost() + cost_diff_) + kThresholdDelta;
+    threshold_ = std::max(pred.sortcost() + cost_diff_, opp_pred.sortcost()) + kThresholdDelta;
```
Comment (genadz) [inline src/thor/bidirectional_astar.cc:761]: as far as we settle edge right after we pop it from the queue, `pred.sortcost() + cost_diff` may be lower than `opp_pred.sortcost()`. so, we should choose bigger in order no to prune search too early

File: test/gurka/test_route.cc [hunk 2]
```diff
@@ -535,3 +535,104 @@ TEST(Standalone, BridgingEdgeIsRestricted) {
...
+  const std::string ascii_map = R"(
+                             F
+                             2
+                             |
+                             E
+                             |
+      A-1--B--------C--------D---------H---------------------------------------I
+                                       |                                       |
+                                       K---------------------------------------J
+  )";
```
Comment (genadz) [inline test/gurka/test_route.cc:551]: special testcase that reproduces bug in master

File: test/gurka/test_route.cc [hunk 3]
```diff
@@ -535,3 +535,104 @@ TEST(Standalone, BridgingEdgeIsRestricted) {
...
+  // This test checks that bidirectional astar doesn't allow double uturns.
+  // It might happens if we handle connection edges incorrectly.
+  const std::string ascii_map = R"(
+  J---------------2---------I-------------------------------F
+                            |
+                            |
+  A---1-------B-------------C------D------------------------E
+              |                    |
+              K--------------------L
+  )";
```
Comment (genadz) [inline test/gurka/test_route.cc:599]: special testcase that reproduce bug in master

File: test/gurka/test_route.cc [hunk 4]
```diff
@@ -535,3 +535,104 @@ TEST(Standalone, BridgingEdgeIsRestricted) {
...
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, nodes, "D", "C")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, nodes, "C", "B")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, nodes, "B", "A")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, nodes, "D", "E")));
+  not_thru_edgeids.push_back(std::get<0>(gurka::findEdgeByNodes(*reader, nodes, "E", "F")));
+
+  test::customize_edges(map.config, [&not_thru_edgeids](const GraphId& edgeid, DirectedEdge& edge) {
+    if (std::find(not_thru_edgeids.begin(), not_thru_edgeids.end(), edgeid) != not_thru_edgeids.end())
+      edge.set_not_thru(true);
+  });
```
Comment (kevinkreiser) [inline test/gurka/test_route.cc:583]: its too bad this didnt happen automatically in the graphenhancer. i suppose you had to do this as a workaround?

**Comment by kevinkreiser** [comment]
@genadz im sorry i have to ask it, hows the performance difference :smile:

File: test/gurka/test_route.cc (hunk 4 above)
Comment (genadz) [reply test/gurka/test_route.cc:583]: right. I needed `not_thru` attributes to reproduce the bug

**Comment by genadz** [comment]
> @genadz im sorry i have to ask it, hows the performance difference

the same as `master`

**Review by kevinkreiser — APPROVED** [review]

**Comment by genadz** [comment]
@kevinkreiser thanks for review!

**Pull request merged**